File: mobile/src/api/interventions.ts
import { apiRequest } from "./client";

export type InterventionResponse = {
  user_id: string;
  week: { start: string; end: string };
  slippage: {
    flagged: boolean;
    reason: string | null;
    completion_rate: number;
    missed_scheduled: number;
  };
  card: {
    title: string;
    message: string;
    options: { key: string; label: string; details: string }[];
  } | null;
  request_id: string;
};

type LatestResult = {
  intervention: InterventionResponse | null;
  requestId: string | null;
  notFound: boolean;
};

export async function getInterventionsLatest(userId: string): Promise<LatestResult> {
  try {
    const { data, response } = await apiRequest<InterventionResponse>(`/interventions/latest?user_id=${userId}`);
    return { intervention: data, requestId: data.request_id || response.headers.get("X-Request-Id"), notFound: false };
  } catch (error) {
    if (error instanceof Error && /404|not found|no intervention snapshot/i.test(error.message)) {
      return { intervention: null, requestId: null, notFound: true };
    }
    throw error;
  }
}

export async function runInterventions(userId: string): Promise<{ intervention: InterventionResponse; requestId: string | null }> {
  const { data, response } = await apiRequest<InterventionResponse>("/interventions/run", {
    method: "POST",
    body: { user_id: userId },
  });
  return { intervention: data, requestId: data.request_id || response.headers.get("X-Request-Id") };
}

File: backend/app/api/schemas/interventions.py
"""Schemas for intervention preview endpoint."""
from __future__ import annotations

from datetime import date
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel

class InterventionRunRequest(BaseModel):
    user_id: UUID

class WeekWindowPayload(BaseModel):
    start: date
    end: date


class SlippagePayload(BaseModel):
    flagged: bool
    reason: Optional[str]
    completion_rate: float
    missed_scheduled: int


class InterventionOption(BaseModel):
    key: str
    label: str
    details: str


class InterventionCard(BaseModel):
    title: str
    message: str
    options: List[InterventionOption]


class InterventionPreviewResponse(BaseModel):
    user_id: UUID
    week: WeekWindowPayload
    slippage: SlippagePayload
    card: Optional[InterventionCard] = None
    request_id: str
